File: src/main/java/com/codingforcookies/betterrecords/common/recipe/RecipeMultiRecord.java
package com.codingforcookies.betterrecords.common.recipe;

import com.codingforcookies.betterrecords.common.BetterRecords;
import com.codingforcookies.betterrecords.common.item.ItemURLRecord;
import com.codingforcookies.betterrecords.common.item.ModItems;
import net.minecraft.inventory.InventoryCrafting;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.IRecipe;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.world.World;

import java.util.ArrayList;

public class RecipeMultiRecord implements IRecipe {

    @Override
    public boolean matches(InventoryCrafting par1InventoryCrafting, World par2World) {
        ArrayList<ItemStack> records = new ArrayList<ItemStack>();

        for(int i = 0; i < par1InventoryCrafting.getSizeInventory(); ++i) {
            ItemStack itemstack = par1InventoryCrafting.getStackInSlot(i);
            if(itemstack != null)
                if(itemstack.getItem() instanceof ItemURLRecord && itemstack.getTagCompound() != null && itemstack.getTagCompound().hasKey("name"))
                    records.add(itemstack);
                else
                    return false;
        }

        return !records.isEmpty() && records.size() != 1;
    }

    @Override
    public ItemStack getCraftingResult(InventoryCrafting par1InventoryCrafting) {
        ArrayList<ItemStack> records = new ArrayList<ItemStack>();

        for(int k = 0; k < par1InventoryCrafting.getSizeInventory(); ++k) {
            ItemStack itemstack = par1InventoryCrafting.getStackInSlot(k);
            if(itemstack != null) {
                if(itemstack.getItem() instanceof ItemURLRecord && itemstack.getTagCompound() != null && itemstack.getTagCompound().hasKey("name"))
                    records.add(itemstack);
                else
                    return null;
            }
        }

        if(records.isEmpty() || records.size() == 1)
            return null;
        else{
            ItemStack itemMultiRecord = new ItemStack(ModItems.itemURLMultiRecord);
            itemMultiRecord.setTagCompound(new NBTTagCompound());

            NBTTagList songs = new NBTTagList();

            for(ItemStack record : records) {
                NBTTagCompound song = new NBTTagCompound();
                song.setString("name", record.getTagCompound().getString("name"));
                song.setString("url", record.getTagCompound().getString("url"));
                song.setString("local", record.getTagCompound().getString("local"));
                songs.appendTag(song);
            }

            itemMultiRecord.getTagCompound().setTag("songs", songs);
            return itemMultiRecord;
        }
    }

    @Override
    public int getRecipeSize() {
        return 10;
    }

    @Override
    public ItemStack getRecipeOutput() {
        return new ItemStack(ModItems.itemURLRecord);
    }

    @Override
    public ItemStack[] getRemainingItems(InventoryCrafting inv) {
        inv.clear();
        return new ItemStack[] {};
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/crafting/recipe/RecipeMultiRecord.kt
package com.codingforcookies.betterrecords.crafting.recipe

import com.codingforcookies.betterrecords.item.ItemRecord
import com.codingforcookies.betterrecords.item.ModItems
import net.minecraft.inventory.InventoryCrafting
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.IRecipe
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraft.util.NonNullList
import net.minecraft.world.World
import net.minecraftforge.registries.IForgeRegistryEntry

import java.util.ArrayList

class RecipeMultiRecord : IForgeRegistryEntry.Impl<IRecipe>(), IRecipe {

    init {
        setRegistryName("recipemultirecord")
    }

    override fun canFit(width: Int, height: Int) = width * height >= 2

    override fun matches(inventoryCrafting: InventoryCrafting, worldIn: World?): Boolean {
        var count = 0

        (0 until inventoryCrafting.sizeInventory)
                .asSequence()
                .map { inventoryCrafting.getStackInSlot(it) }
                .filter { !it.isEmpty }
                .forEach {
                    if (it.item is ItemRecord && it.hasTagCompound() && it.tagCompound!!.hasKey("name")) {
                        count++
                    } else {
                        return false
                    }
                }

        return count > 1
    }

    override fun getCraftingResult(inventoryCrafting: InventoryCrafting): ItemStack? {
        val records = ArrayList<ItemStack>()

        (0 until inventoryCrafting.sizeInventory)
                .map { inventoryCrafting.getStackInSlot(it) }
                .filter { !it.isEmpty }
                .forEach {
                    records.add(it)
                }

        return ItemStack(ModItems.itemMultiRecord).apply {
            tagCompound = NBTTagCompound()
            val songs = NBTTagList()

            records.forEach {
                songs.appendTag(NBTTagCompound().apply {
                    setString("name", it.tagCompound!!.getString("name"))
                    setString("url", it.tagCompound!!.getString("url"))
                    setString("local", it.tagCompound!!.getString("local"))
                })
            }

            tagCompound!!.setTag("songs", songs)
        }
    }

    override fun getRecipeOutput() = ItemStack(ModItems.itemRecord)

    override fun getRemainingItems(inv: InventoryCrafting): NonNullList<ItemStack> {
        inv.clear()
        return NonNullList.create()
    }
}
